### PR TITLE
Parse tree for `impl` that is better for check stage

### PR DESCRIPTION
**Pull request by josh11b**
Use two different nodes for "<type> followed by `as`" and "<type> omitted before `as`, use `self`", so it is easier to determine which case. Later the second case will push the type id for `self` onto the node stack, making the two paths more similar.

### CODE REVIEW

File: toolchain/parse/node_kind.def [hunk 1]
```diff
@@ -683,7 +683,14 @@ CARBON_PARSE_NODE_KIND_CHILD_COUNT(ImplIntroducer, 0, Impl)
...
-CARBON_PARSE_NODE_KIND_CHILD_COUNT(ImplAs, 0, As)
+
+// `... as`:
+//   _external_: expression
+// TypeImplAs
+CARBON_PARSE_NODE_KIND_CHILD_COUNT(TypeImplAs, 1, As)
+// `as` without a type before it
+CARBON_PARSE_NODE_KIND_CHILD_COUNT(DefaultSelfImplAs, 0, As)
+
 CARBON_PARSE_NODE_KIND_BRACKET(ImplDefinitionStart, ImplIntroducer,
```
Comment (jonmeow) [inline toolchain/parse/node_kind.def]: Should these parse nodes be more clearly associated with the comment that talks about them? Particularly as more is in the middle, I also wonder if the usual "external" would be better than "see below"

File: toolchain/parse/typed_nodes.h [hunk 2]
```diff
@@ -859,6 +857,16 @@ struct ImplForall {
   ImplicitParamListId params;
 };
 
+// `as` with no type before it
```
Comment (jonmeow) [inline toolchain/parse/typed_nodes.h:860]:
```suggestion
// `as` with no type before it.
```

Similarly, this feels more like a sentence with the additional words, so punctuation might make sense.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

File: toolchain/check/handle_impl.cpp [hunk 3]
```diff
@@ -32,20 +32,24 @@ auto HandleImplForall(Context& context, Parse::ImplForallId parse_node)
...
     -> bool {
+  auto self_id = context.node_stack().PopExpr();
+  context.node_stack().Push(parse_node, self_id);
+  return true;
+}
+
+auto HandleDefaultSelfImplAs(Context& /*context*/,
+                             Parse::DefaultSelfImplAsId /*parse_node*/)
+    -> bool {
+  // TODO: Determine self_id and push it onto node stack
```
Comment (jonmeow) [inline toolchain/check/handle_impl.cpp]:
```suggestion
  // TODO: Determine self_id and push it onto node stack.
```

Punctuation: https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

**Review by jonmeow — APPROVED** [review]
LGTM. Comments are minor, feel free to merge regardless of whether you make the suggested changes.

File: toolchain/parse/handle_impl.cpp [hunk 4]
```diff
@@ -64,17 +63,16 @@ auto HandleImplAfterForall(Carbon::Parse::Context& context) -> void {
...
 auto HandleImplBeforeAs(Carbon::Parse::Context& context) -> void {
   auto state = context.PopState();
-  if (state.has_error) {
-    context.ReturnErrorOnState();
-    return;
-  }
   if (auto as = context.ConsumeIf(Lex::TokenKind::As)) {
-    context.AddLeafNode(NodeKind::ImplAs, *as);
+    context.AddNode(NodeKind::TypeImplAs, *as, state.subtree_start,
+                    state.has_error);
```
Comment (zygoloid) [inline toolchain/parse/handle_impl.cpp:68]: It looks like we don't have any test coverage for the case where there's a parse error before `as`, where we'd be building a subtree with `has_error == true` here. I *think* this is handling that properly, but it'd be good to have a test case to cover that.

File: toolchain/parse/typed_nodes.h (hunk 2 above)
Comment (josh11b) [reply toolchain/parse/typed_nodes.h:860]: This is a noun phrase not a sentence, and adding a period would be inconsistent with the parallel comments

File: toolchain/parse/handle_impl.cpp (hunk 4 above)
Comment (josh11b) [reply toolchain/parse/handle_impl.cpp:68]: Added.

File: toolchain/parse/node_kind.def (hunk 1 above)
Comment (josh11b) [reply toolchain/parse/node_kind.def]: Done.

**Pull request merged**